### ISSUE TITLE
Flycheck corrupts buffer when the underlying checker issues some unhandled warnings

### ISSUE THREAD

**Issue by chihyang** [flycheck/flycheck#1848]
Thank you for taking the time to report an issue and improve Flycheck.  This template is for **actual bugs you observed**.  If you have trouble setting up Flycheck, or if you have a question, please use the relevant issue template instead.

## Checklist
- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description
Racket uses [`raco`](https://docs.racket-lang.org/raco/) as its command line tool. As the flycheck [doc](https://www.flycheck.org/en/28/_downloads/flycheck.html#Racket) says, flycheck uses `raco` for error checking. `raco` is also used as the package manager for racket. I customize some racket packages for debugging. Thus I have to install a racket package twice: one is system wide and installed together with racket; the other is local under my home directory. In this case, when `raco` is running, it will use the local version of the package, but gives a warning like:

``` shell
warning: tool "scribble" registered twice: "scribble" and #<path:/usr/share/racket/pkgs/scribble-lib/scribble>
```

Don't worry, this warning is harmless. Everything just works fine.

As the screenshot below shows, when I install two versions of the same racket package and edit a racket file in emacs with `flycheck-mode` enabled, flycheck keeps writing the harmless warning of `raco` into my file. I did some debugging. I guess this is because the first writing causes file change and syntax error, which triggers flycheck to call `raco` again, leading to an infinite loop. It seems flycheck does some redirection or buffer writing. Things that should have been written into standard error are written to my file.

## Steps to reproduce
1. Make some changes to a system wide racket package and install it in user directory (more detailed steps can be provided if necessary)
2. run `raco`(no argument needed), just to make sure you see something like `warning: tool "xxx" registered twice: xxx`
3. Open a racket file in emacs
4. Enable `flycheck-mode`
5. See screenshot

- [x] I have read https://emacs.stackexchange.com/questions/28429/how-do-i-troubleshoot-emacs-problems

## Expected behavior
Even if the underlying checker reports a warning, the checked buffer should keep untouched.

## Screenshots
![flycheck_corrupts_file](https://user-images.githubusercontent.com/3126902/102487260-394c1100-40a5-11eb-83d9-566c335baad0.gif)

## System configuration
```
Syntax checkers for buffer style.rkt in racket-mode:

First checker to run:

  racket
    - may enable:            yes
    - may run:               t
    - executable:            Found at /usr/bin/raco
    - compiler-lib package:  present
    - Geiser Implementation: Using Racket Mode

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled checkers.

--------------------

Flycheck version: 32snapshot (package: 20201214.2154)
Emacs version:    27.1
System:           x86_64-pc-linux-gnu
Window system:    x
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration
- [ ] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

## Additional notes
This problem may not limit to `raco`. Just in normal case, it's not that easy to happen.

[Troubleshooting guide]:  https://www.flycheck.org/en/latest/user/troubleshooting.html

**Comment by chihyang** [comment]
It seems the following code causes the problem:

https://github.com/flycheck/flycheck/blob/a11b789807d1d942d6fcfac17508d072b9cf7ba8/flycheck.el#L10756-L10758

Because flycheck calls `call-process`, and `call-process` will insert standard error into current buffer if DESTINATION is set to `t` (see the [doc](https://www.gnu.org/software/emacs/manual/html_node/elisp/Synchronous-Processes.html)), call to `flycheck-racket-has-expand-p` causes buffer change. I change it to 

``` emacs-lisp
(defun flycheck-racket-has-expand-p (checker)
  "Whether the executable of CHECKER provides the `expand' command."
  (eql 0 (flycheck-call-checker-process checker nil '(t nil) nil "expand")))
```

to discard standard error output. Now my buffer won't be changed. Do you think this change is okay?

**Comment by cpitclaudel** [comment]
Hi @chihyang !  Thanks a lot for the report, and further thanks for diagnosing the issue.  Shouldn't it simply be the following, though?  It doesn't seem that we use stdout either, and it would go to the buffer as well:

```
(defun flycheck-racket-has-expand-p (checker)
  "Whether the executable of CHECKER provides the `expand' command."
  (eql 0 (flycheck-call-checker-process checker nil nil nil "expand")))
```

If you can prepare a patch, that would be wonderful.

**Comment by chihyang** [comment]
Okay, see PR #1849.
